### PR TITLE
Explain integration sprint 28

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/util/MathUtils.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    private static final int NUMBER_OF_ESTIMATES = 50;
+    private static final int NUMBER_OF_PROBES = 100;
+    private static final int NUMBER_OF_VALUES = 500;
+    private static final double[] COLLISION_ESTIMATES = new double[NUMBER_OF_ESTIMATES + 1];
+
+    static {
+        generateEstimates();
+    }
+
+    public static double estimateNumberOfHashCollisions(int numberOfValues, int hashSize)
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/util/MathUtils.java:29]: I don't see tests for this

File: presto-main/src/main/java/com/facebook/presto/util/MathUtils.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    {
+        if (hashSize == 0) {
+            return 0d;
+        }
+
+        return COLLISION_ESTIMATES[NUMBER_OF_ESTIMATES * numberOfValues / hashSize] * numberOfValues / NUMBER_OF_VALUES;
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/util/MathUtils.java:35]: You could easily improve estimation quality by extrapolation of two closest values - it's extremely simple:

```
double lowerEstimation = COLLISION_ESTIMATES[floor(NUMBER_OF_ESTIMATES * numberOfValues / hashSize)];
double upperEstimation = COLLISION_ESTIMATES[ceil(NUMBER_OF_ESTIMATES * numberOfValues / hashSize)];
double estimationDistanceFromLower = NUMBER_OF_ESTIMATES * numberOfValues / hashSize - floor(NUMBER_OF_ESTIMATES * numberOfValues / hashSize);
double estimation = lowerEstimation * estimationDistanceFromLower + upperEstimation * (1 - estimationDistanceFromLower)
return estimation * numberOfValues / NUMBER_OF_VALUES;
```

File: presto-main/src/main/java/com/facebook/presto/util/MathUtils.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ */
+package com.facebook.presto.util;
+
+import java.util.Random;
+
+public final class MathUtils
+{
+    private static final int NUMBER_OF_ESTIMATES = 50;
+    private static final int NUMBER_OF_PROBES = 100;
+    private static final int NUMBER_OF_VALUES = 500;
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/util/MathUtils.java:22]: Is the 500 enough?

File: presto-main/src/main/java/com/facebook/presto/util/MathUtils.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    static {
+        generateEstimates();
+    }
+
+    public static double estimateNumberOfHashCollisions(int numberOfValues, int hashSize)
+    {
+        if (hashSize == 0) {
+            return 0d;
+        }
+
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/util/MathUtils.java:34]: `checkState(0 <= numberOfValues <= hashSize);`  ?

File: presto-main/src/main/java/com/facebook/presto/util/MathUtils.java (hunk 2 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/util/MathUtils.java:35]: Of course it makes sense to interpolate results. However, the posted snippet is wrong.

File: presto-main/src/main/java/com/facebook/presto/util/MathUtils.java (hunk 3 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/util/MathUtils.java]: hard to say, I will increase it to 5000

Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/util/MathUtils.java:22]: I will actually leave it 500, changing it to 5000 increases app startup time

**Comment by sopel39** [comment]
applied comments

**Pull request closed without merging**